python/pyspark/sql/connect/column.py

```diff
@@ -352,17 +365,326 @@ def __init__(self, expr: Expression) -> None:
+        return _bin_op("like")(self, other)
+
+    def ilike(self: "Column", other: str) -> "Column":
+        """
+        SQL ILIKE expression (case insensitive LIKE). Returns a boolean :class:`Column`
+        based on a case insensitive match.
+
+        Parameters
+        ----------
+        other : str
+            a SQL LIKE pattern
+        See Also
+        --------
+        pyspark.sql.Column.rlike
+        Returns
+        -------
+        :class:`Column`
+            Column of booleans showing whether each element
+            in the Column is matched by SQL LIKE pattern.
+
+        Examples
+        --------
+        >>> df = spark.createDataFrame(
+        ...      [(2, "Alice"), (5, "Bob")], ["age", "name"])
+        >>> df.filter(df.name.ilike('%Ice')).collect()
+        [Row(age=2, name='Alice')]
+        """
+        return _bin_op("ilike")(self, other)
+
+    @overload
+    def substr(self, startPos: int, length: int) -> "Column":
+        ...
+
+    @overload
+    def substr(self, startPos: "Column", length: "Column") -> "Column":
+        ...
+
+    def substr(self, startPos: Union[int, "Column"], length: Union[int, "Column"]) -> "Column":
+        """
+        Return a :class:`Column` which is a substring of the column.
+
+        Parameters
+        ----------
+        startPos : :class:`Column` or int
+            start position
+        length : :class:`Column` or int
+            length of the substring
+        Returns
+        -------
+        :class:`Column`
+            Column representing whether each element of Column is substr of origin Column.
+
+        Examples
+        --------
+        >>> df = spark.createDataFrame(
+        ...      [(2, "Alice"), (5, "Bob")], ["age", "name"])
+        >>> df.select(df.name.substr(1, 3).alias("col")).collect()
+        [Row(col='Ali'), Row(col='Bob')]
+        """
+        if type(startPos) != type(length):
+            raise TypeError(
+                "startPos and length must be the same type. "
+                "Got {startPos_t} and {length_t}, respectively.".format(
+                    startPos_t=type(startPos),
+                    length_t=type(length),
+                )
+            )
+        from pyspark.sql.connect.function_builder import functions as F
+
+        if isinstance(length, int):
+            length_exp = self._lit(length)
+        elif isinstance(length, Column):
+            length_exp = length
+        else:
+            raise TypeError("Unsupported type for substr().")
+
+        if isinstance(startPos, int):
+            start_exp = self._lit(startPos)
+        else:
+            start_exp = startPos
+
+        return F.substr(self, start_exp, length_exp)
 
     def __eq__(self, other: Any) -> "Column":  # type: ignore[override]
         """Returns a binary expression with the current column as the left
         side and the other expression as the right side.
         """
-        from pyspark.sql.connect._typing import PrimitiveType
```

Comment:
I am pretty it does not solve the root problem...

If you can column.py, _typing.py and functions.py, you could see that _typing.py and functions.py uses `Column`, then whenever `column.py` wants to include anything from that two classes, it become circular import

Reply:
normally, we put those things under:
```
if TYPE_CHECKING:
    from pyspark.sql.connect._typing import PrimitiveType
```
then it will be ok for mypy.

The problem here is `get_args(PrimitiveType)` requires `_typing`. Maybe we can just use `(bool, float, int, str)` instead of `get_args(PrimitiveType)`

Reply:
cc @HyukjinKwon

Reply:
I got something like this which actually is the case that code cannot run (so not only fail mypy check) 
```
  File "/Users/rui.wang/Documents/github/spark/python/pyspark/sql/connect/_typing.py", line 29, in <module>
    from pyspark.sql.connect.column import ScalarFunctionExpression, Column
ImportError: cannot import name 'ScalarFunctionExpression' from partially initialized module 'pyspark.sql.connect.column' (most likely due to a circular import) (/Users/rui.wang/Documents/github/spark/python/pyspark/sql/connect/column.py)
```

Reply:
ok, let's fix it as followup